### DISCUSSION TITLE
get rid of OrderedDict

### DISCUSSION THREAD

**Discussion by Boorhin** [OpenDrift/opendrift#1165] [Ideas]
OrderectDict specifics used by Opendrift has been part of python since 3.6 and doesn't need to be part of it anymore. It also slower than the native dictionaries present in python 3 apparently. Collection is also not very supported for that side of things if I understood well.
I have found these occurrences in the code 
https://github.com/search?q=repo%3AOpenDrift%2Fopendrift%20OrderedDict&type=code
In 10 minutes I have removed the imports and the use of it.
OpenDrift works perfectly fine without it and it is an opportunity to remove a library import and a layer of clutter.

**Comment by gauteh** [comment]
What do you mean? What do you replace the OrderedDict with?

**Comment by Boorhin** [comment]
a regular dictionary

**Comment by Boorhin** [comment]
an example

``` python
variables = {
        'ID': {'dtype': np.int32,  # Unique numerical identifier
                'seed': False,
                'default': -1},  # ID to be assigned by application
        'status': {'dtype': np.int32,  # Status categories
                    'seed': False,
                    'default': 0},
        'moving': {'dtype': np.int32,  # Set to 0 for elements which are frosen
                    'seed': False,
                    'default': 1},
        'age_seconds': {'dtype': np.float32,
                         'units': 's',
                         'seed': False,
                         'default': 0},
        'origin_marker': {'dtype': np.int16,
                           'unit': '',
                           'description': '''An integer kept constant during the simulation. 
                             Different values may be used for different seedings, 
                             to separate elements during analysis. 
                             With GUI, only a single seeding is possible.''',
                           'default': 0},
        'lon': {'dtype': np.float32,
                 'units': 'degrees_east',
                 'standard_name': 'longitude',
                 'long_name': 'longitude',
                 'seed': False,
                 'axis': 'X'},
        'lat': {'dtype': np.float32,
                 'units': 'degrees_north',
                 'standard_name': 'latitude',
                 'long_name': 'latitude',
                 'seed': False,
                 'axis': 'Y'},
        'z': {'dtype': np.float32,
                   'units': 'm',
                   'standard_name': 'z',
                   'long_name': 'vertical position',
                   'axis': 'Z',
                   'positive': 'up',
                   'default': 0}
        }
```

**Comment by Boorhin** [comment]
This is the difference the top python process is blocked on a single cpu (4% is 1/24) with OrderedDict whereas the second one using the native (here 3.10) dictionary type that has some parallelisable operations.
![image](https://github.com/OpenDrift/opendrift/assets/9576982/4afa093a-163a-4062-8ad6-82d955260c38)

**Comment by gauteh** [comment]
Ok, some places can probably be replaced with regular dicts. Like the one you suggest. But many other places require the order, like priority_readers.

**Reply by gauteh** [reply]
Now I understand. If all dicts are ordered this should be a safe change. Note that we are currently working on a big change in https://github.com/OpenDrift/opendrift/pull/1164, so might be better to base your changes on that one.

**Reply by Boorhin** [reply]
I can describe the changes and you can implement them yourself once it is time. It really takes 10 min max. I struggle to submit new changes as my fork doesn't pass your tests and I must admit I am not too sure why as it is perfectly functional. 
Actually I think that `variables` should be a xarray Dataset. That would really make the code easier to handle and more square but I haven't been bothered to make sure of the compatibility. But that's for another time.
